sharedbuffers/mapped_struct.py

```diff
@@ -317,113 +323,331 @@ def unpack_from(cls, buf, offs, idmap = None):
-        return mapped_tuple.pack_into(obj, buf, offs, idmap, implicit_offs)
+        elif dcode in ('q', 'd', 't'):
+            objlen = cython.cast(cython.p_longlong, pbuf + dataoffs)[0] >> 8
+            dataoffs += 8
+            return dcode, objlen, 8, dataoffs
 
-    @classmethod
-    def unpack_from(cls, buf, offs, idmap = None, array = array.array, itemsizes = {
-                dtype : array.array(dtype, []).itemsize
-                for dtype in ('B','b','H','h','I','i','l','d')
-            } ):
-        if idmap is None:
-            idmap = {}
-        if offs in idmap:
-            return idmap[offs]
+        else:
+            raise ValueError("Inconsistent data, unknown type code %r" % (dcode,))
 
-        buf = _likerobuffer(buf)
-        dcode = buf[offs]
-        dataoffs = offs
+    def _metadata(self,
+        itemsizes =  {dtype : array.array(dtype, []).itemsize for dtype in ('B','b','H','h','I','i','l','d')},
+        ):
+
+        dataoffs = self.offs
+        buf = self.buf
+        dcode = buf[dataoffs]
 
         if dcode in ('B','b','H','h','I','i'):
             objlen, = struct.unpack('<I', buf[dataoffs+1:dataoffs+4] + '\x00')
             dataoffs += 4
             if objlen == 0xFFFFFF:
                 objlen = struct.unpack_from('<Q', buf, dataoffs)
                 dataoffs += 8
-            rv = cls(buf, offs, dataoffs, dcode, itemsizes[dcode], objlen, idmap)
+            return dcode, objlen, itemsizes[dcode], dataoffs, struct.Struct(dcode)
+
         elif dcode == 'q':
             objlen, = struct.unpack('<Q', buf[dataoffs+1:dataoffs+8] + '\x00')
             dataoffs += 8
-            rv = cls(buf, offs, dataoffs, dcode, itemsizes['l'], objlen, idmap)
+            return dcode, objlen, itemsizes['l'], dataoffs, struct.Struct('l')
+
         elif dcode == 'd':
             objlen, = struct.unpack('<Q', buf[dataoffs+1:dataoffs+8] + '\x00')
             dataoffs += 8
-            rv = cls(buf, offs, dataoffs, dcode, itemsizes[dcode], objlen, idmap)
+            return dcode, objlen, itemsizes['d'], dataoffs, struct.Struct('d')
+
         elif dcode == 't':
             objlen, = struct.unpack('<Q', buf[dataoffs+1:dataoffs+8] + '\x00')
             dataoffs += 8
-            rv = cls(buf, offs, dataoffs, dcode, itemsizes['l'], objlen, idmap)
+            return dcode, objlen, itemsizes['l'], dataoffs, struct.Struct('l')
+
         else:
             raise ValueError("Inconsistent data, unknown type code %r" % (dcode,))
-        return rv
 
-    @cython.locals(obj_offs = cython.ulonglong, dcode = object, index=cython.ulonglong)
+    def __init__(self, buf, offs, idmap = None):
+        self.offs = offs
+        self.buf = buf
+        self.idmap = {} if idmap is None else idmap
+
+        if cython.compiled:
+            self._cinit()
+            self._cmetadata()
+        else:
+            # Call metadata to check the object
+            self._metadata()
```

Comment:
Why different methods?

Why can't you just have the `if cython.compiled` in _metadata?

Reply:
Because these methods are long and I think that it is much clear in this way.